### ISSUE TITLE
borg crash during prune, check and create: KeyError in locking.py, and "AssertionError: bug in code, exclusive lock should exist here"

### ISSUE THREAD

**Issue by ress** [borgbackup/borg#3590]
My storage server's hard disk ran out of space a few days ago, and after removing some placeholder files and pruning, I was able to make backups again. Then, during the first backup, I ran out of space again and wanted to prune more aggressively.

That second prune, and every other `prune`, `check`, or `create` I've tried failed with errors below. I've tried `borg check --repair` as well, and a similar error appeared. I'm running these commands directly on my repository server, but the initial backups and the first prune happened via the remote repository syntax in borg.

What I've tried so far:
* Resetting any locks with `break-lock` multiple times, but I don't notice a difference in the error messages.
* Scrubbing the underlying ZFS file system of the repo server. No errors.
* I've read #3082, but couldn't find anything helpful for my situation in there.

At this point I still assume it's something I've done wrong, but the "AssertionError: bug in code, exclusive lock should exist here" error looks suspicious. Is there a way to recover my repo, or do I need to start fresh?

My system:
* borg 1.1.4 (the static build, linux64 and freebsd64)
* FreeBSD (repo server): `FreeBSD laniakea 11.1-RELEASE-p4 FreeBSD 11.1-RELEASE-p4 #0: Tue Nov 14 06:12:40 UTC 2017     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64`
* Linux (system making backups): `Linux virgo 4.8.0-36-generic #36~16.04.1-Ubuntu SMP Sun Feb 5 09:39:57 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

**borg prune crash:**
```
[locutus@laniakea /home/locutus]$ borg prune -s --keep-within=10d --keep-monthly=12 --list --save-space -v virgo-backup/
Enter passphrase for key /locutus/virgo-backup:
Keeping archive: virgo-2018-01-31-06:00:01.checkpoint Wed, 2018-01-31 06:00:05 [4ff5c8c4091fcbcded91b9e6ba6c858d1a8072773bc35a24235d792516e6cdf4]
Pruning archive: virgo-2018-01-26-06:00:01.checkpoint Fri, 2018-01-26 06:00:05 [ebd7844d5161c4f7bb102f8419f0f129b41b88a25e336028e8fe8258c0a01234] (1/6)
Exception ignored in: <bound method Repository.__del__ of <Repository /locutus/virgo-backup>>
Traceback (most recent call last):
  File "borg/repository.py", line 170, in __del__
  File "borg/repository.py", line 408, in close
  File "borg/locking.py", line 379, in release
  File "borg/locking.py", line 278, in modify
KeyError: (('laniakea@345040274363', 5333, 0),)
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 150, in wrapper
  File "borg/archiver.py", line 1562, in do_prune
  File "borg/archive.py", line 791, in delete
  File "borg/archive.py", line 774, in chunk_decref
  File "borg/cache.py", line 908, in chunk_decref
  File "borg/repository.py", line 1082, in delete
  File "borg/repository.py", line 466, in prepare_txn
AssertionError: bug in code, exclusive lock should exist here

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "borg/archiver.py", line 4175, in main
  File "borg/archiver.py", line 4107, in run
  File "borg/archiver.py", line 150, in wrapper
  File "borg/repository.py", line 195, in __exit__
  File "borg/repository.py", line 408, in close
  File "borg/locking.py", line 379, in release
  File "borg/locking.py", line 278, in modify
KeyError: ('laniakea@345040274363', 5333, 0)

Platform: FreeBSD laniakea 11.1-RELEASE-p4 FreeBSD 11.1-RELEASE-p4 #0: Tue Nov 14 06:12:40 UTC 2017     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC amd64 amd64
Borg: 1.1.4  Python: CPython 3.5.4
PID: 5333  CWD: /locutus
sys.argv: ['borg', 'prune', '-s', '--keep-within=10d', '--keep-monthly=12', '--list', '--save-space', '-v', 'virgo-backup/']
SSH_ORIGINAL_COMMAND: None

[locutus@laniakea /home/locutus]$
```

**borg check crash:**

```
[locutus@laniakea /home/locutus]$ borg check -v virgo-backup
Starting repository index check
Completed repository check, no problems found.
Starting archive consistency check...
Enter passphrase for key /locutus/virgo-backup:
Analyzing archive virgo-2017-03-24-01:54:49 (1/18)
Exception ignored in: <bound method Repository.__del__ of <Repository /locutus/virgo-backup>>
Traceback (most recent call last):
  File "borg/repository.py", line 170, in __del__
  File "borg/repository.py", line 408, in close
  File "borg/locking.py", line 379, in release
  File "borg/locking.py", line 278, in modify
KeyError: (('laniakea@345040274363', 87648, 0),)
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 150, in wrapper
  File "borg/archiver.py", line 311, in do_check
  File "borg/archive.py", line 1180, in check
  File "borg/archive.py", line 1563, in rebuild_refcounts
  File "borg/archive.py", line 233, in add
  File "borg/archive.py", line 250, in flush
  File "borg/archive.py", line 1367, in add_callback
  File "borg/crypto/key.py", line 367, in encrypt
  File "borg/crypto/nonces.py", line 72, in ensure_reservation
  File "borg/crypto/nonces.py", line 39, in get_repo_free_nonce
  File "borg/repository.py", line 303, in get_free_nonce
AssertionError: bug in code, exclusive lock should exist here

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "borg/archiver.py", line 4175, in main
  File "borg/archiver.py", line 4107, in run
  File "borg/archiver.py", line 150, in wrapper
  File "borg/repository.py", line 195, in __exit__
  File "borg/repository.py", line 408, in close
  File "borg/locking.py", line 379, in release
  File "borg/locking.py", line 278, in modify
KeyError: ('laniakea@345040274363', 87648, 0)

Platform: FreeBSD laniakea 11.1-RELEASE-p4 FreeBSD 11.1-RELEASE-p4 #0: Tue Nov 14 06:12:40 UTC 2017     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC amd64 amd64
Borg: 1.1.4  Python: CPython 3.5.4
PID: 87648  CWD: /locutus
sys.argv: ['borg', 'check', '-v', 'virgo-backup/']
SSH_ORIGINAL_COMMAND: None

[locutus@laniakea /home/locutus]$
```

**borg create crash** (on linux backup client)
```
root@virgo:~# ./backup.sh
------------------------------------------------------------------------------
Archive name: virgo-2018-02-04-10:29:10
Archive fingerprint: 36c2eae67c58c8124278bb8ae76e9cb8bb3ba1f720ef785ee5a94ec28b9d35de
Time (start): Sun, 2018-02-04 10:29:13
Time (end):   Sun, 2018-02-04 10:31:33
Duration: 2 minutes 19.79 seconds
Number of files: 53342
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:              206.99 GB            204.41 GB              6.45 MB
All archives:               36.75 TB             30.28 TB              2.85 TB

                       Unique chunks         Total chunks
Chunk index:                 2007635             27322743
------------------------------------------------------------------------------
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 4175, in main
  File "borg/archiver.py", line 4107, in run
  File "borg/archiver.py", line 150, in wrapper
  File "borg/archiver.py", line 545, in do_create
  File "borg/cache.py", line 469, in __exit__
  File "borg/cache.py", line 498, in close
  File "borg/cache.py", line 313, in close
  File "borg/locking.py", line 379, in release
  File "borg/locking.py", line 278, in modify
KeyError: ('virgo@26830026377318', 16740, 0)

Platform: Linux virgo 4.8.0-36-generic #36~16.04.1-Ubuntu SMP Sun Feb 5 09:39:57 UTC 2017 x86_64 x86_64
Linux: debian stretch/sid
Borg: 1.1.4  Python: CPython 3.5.4
PID: 16740  CWD: /root
sys.argv: ['/usr/local/bin/borg', 'create', '-C', 'zlib,7', '-s', '-p', '--exclude-from', '/root/excludes.txt', '-v', 'locutus@laniakea:virgo-backup::virgo-2018-02-04-10:29:10', '/triangle/Photos']
SSH_ORIGINAL_COMMAND: None
```

**Comment by ThomasWaldmann** [comment]
Notes:
- reported PIDs in traceback consistent with PIDs in locking, but still `KeyError`
- do_prune does exclusive=True on the RemoteRepository

**Comment by ress** [comment]
I think I figured out what caused these errors, and it looks like user error after all. Despite what I thought, my systemd timer that triggers backups and prunes the repo was _not_ turned off and regularly started up, and then failed shortly after. I've configured everything so that the script would restart 5 seconds later, attempting to break-lock & rerun the backup/prune command.

This meant that two borg processes tried accessing the same repo at the same time, which looks like it always failed eventually (a few seconds to minutes after the second process started) – probably after the failed backup/prune job launched by systemd destroyed the lock, or the backup creation ran out of disk space on the repository server.

Now that I really turned off the systemd timer, all actions on the repository work as expected again.

Thanks for your notes from a few days ago – thinking about locks and how they might disappear led me to check all processes that could access the repository again, which turned up the still-enabled systemd timer.
